src/audio/rtnr/rtnr.c

```diff
@@ -396,17 +479,149 @@ static int rtnr_cmd_get_data(struct comp_dev *dev,
 	return ret;
 }
 
-static int rtnr_cmd_set_data(struct comp_dev *dev,
+static int rtnr_reconfigure(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint8_t *config;
+	size_t size;
+
+	comp_dbg(dev, "rtnr_reconfigure()");
+
+	config = comp_get_data_blob(cd->model_handler, &size, NULL);
+	comp_dbg(dev, "rtnr_reconfigure() size: %d", size);
+
+	if (size == 0) {
+		/* No data to be handled */
+		return 0;
+	}
+
+	if (!config) {
+		comp_err(dev, "rtnr_reconfigure(): Config not set");
+		return -EINVAL;
+	}
+
+	comp_info(dev, "rtnr_reconfigure(): New data applied %p (%zu bytes)",
+		  config, size);
+
+	cd->reconfigure = false;
+
+	RTKMA_API_Set(cd->rtk_agl, config, size, RTNR_DATA_ID_PRESET);
+
+	return 0;
+}
+
+static int rtnr_set_comp_config_by_ipc_config_process(struct comp_dev *dev,
+							struct ipc_config_process *config)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int ret;
+
+	comp_dbg(dev, "rtnr_set_comp_config_by_ipc_config(): %d", config->size);
+	if (config->size != sizeof(cd->config)) {
+		comp_err(dev, "rtnr_set_comp_config_by_ipc_config(): invalid size %d",
+			config->size);
+		return -EINVAL;
+	}
+
+	ret = memcpy_s(&cd->config,
+				sizeof(cd->config),
+				config->data,
+				config->size);
+	if (ret)
+		return ret;
+
+	comp_info(dev,
+		"rtnr_set_comp_config_by_ipc_config(): sample_rate = %d, enabled=%d",
+		cd->config.params.sample_rate,
+		cd->config.params.enabled);
+
+	return ret;
+}
+
+static int rtnr_set_comp_config_by_ipc_ctrl_data(struct comp_dev *dev,
 							struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int ret;
+
+	comp_dbg(dev, "rtnr_set_comp_config_by_ipc_ctrl_data(): size: %d",
+		cdata->data->size);
+
+	if (cdata->data->size != sizeof(cd->config)) {
+		comp_err(dev,
+			"rtnr_set_comp_config_by_ipc_ctrl_data(): invalid size %d",
+			cdata->data->size);
+
+		return -EINVAL;
+	}
+
+	ret = memcpy_s(&cd->config,
+				sizeof(cd->config),
+				cdata->data->data,
+				cdata->data->size);
+	if (ret)
+		return ret;
+
+	comp_info(dev,
+		"rtnr_set_comp_config_by_ipc_ctrl_data(): sample_rate = %d, enabled=%d",
+		cd->config.params.sample_rate,
+		cd->config.params.enabled);
+
+	return ret;
+}
+
+static int rtnr_set_bin_data(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
 {
 	struct comp_data *cd = comp_get_drvdata(dev);
 	int ret = 0;
 
+	assert(cd);
```

Comment:
in some handlers, called from `rtnr_cmd()` `cd` is used with no check, in `rtnr_get_bin_data()` you return an error and here it's an `assert()`. Shouldn't that be handled uniformly? Maybe check for it in `rtnr_cmd()` and return an error immediately?